FIX: ISX-1890 + Improved asset management in Input Action Asset editors

ekcoh opened the pull request:
### Description

This branch introduces improved asset management of .inputactions files open in editors and fixes a bug where pressing Save Asset in Input Action Editor window would throw an exception if the file was moved while the editor was being open. For details on changes introduced see Changes made below.

### Changes made

The following changes have been made to this branch:
- Fixed an issue where deleting an `.inputactions` asset with the asset open for editing in a Input Action Editor window (UITK) would cause the editor to crash with `FileNotFoundException`.
- Fixed an issue where UITK and IMGUI editors have different behaviours when moving an asset with editor open at the same time. (Move/drag from within Unity + external file system rename). 
- Fixed an issue where UITK and IMGUI editors have different behaviours when deleting an asset with editor open at the same time. (Delete from within Unity + external file system delete)
- Fixed an issue where UITK and IMGUI editors have different behaviours when renaming an asset with editor open at the same time. (Rename from within Unity + external file system rename)
- Refactored and added same AssetModificationProcessor and AssetPostProcessor monitors to both IMGUI and UITK editors to avoid overlapping functionality. This is implemented as `InputActionAssetPostprocessor` and `InputActionAssetModificationProcessor` replacing existing processor in IMGUI and introducing it for UITK editor.
- Refactored and moved all dialog logic into Dialogs.cs and converted all outcomes to enum values for increased readability. Also added a layer of abstraction to allow mocking and "head-less" test execution.
- Modified UITK dispatch mechanism to show full exception traces in log instead of truncating tracing information to messages only to simplify tracking errors and bugs.
- Modified `InputActionsEditorState` to rely on GUID-to-index and index-to-GUID remapping when reimporting an asset and it is different from previous incarnation of the asset. Since renaming, restructuring, deleting and additions may happen to any map, action, binding, interaction, processor or control scheme during reimport it makes sense to rely on GUIDs for identification rather than names since not stable identifiers.
- Fixed an issue where the UITK Input Action Editor would show incorrect modification state if open during domain reload.

Test related changes:
- Added `AssetDatabaseUtils.cs` which simplifies writing test cases involving assets since it removes the burden from the developer to find unique file and directory names as well as automatically cleans-up all associated files and folders as part of InputTestFixture teardown.
- Corrected incorrect one-time test fixture setup and teardown in `CoreTest_ProjectWideActions.cs` which resulted in backup folder being left in /Assets after running tests.
- Fixed test case `Editor_InputActionAssetManager_CanMoveAssetOnDisk `which was testing InputActionAssetManager monitoring of external file system changes but failed to move .meta files along with assets generating warnings in test runner.
- Expanded test coverage of InputActionAssetManager via tests called out in checklist below. This was done to support refactoring of the class with less worry about breaking existing behavior without improving it. Also moved those tests into its own .cs file targeting this specific class.
- Updated `InputTestFixture` to apply default mocking and restore for all editor test cases to avoid running into problems with user dialogs affecting execution flow.

### Notes

Project-wide settings editor still operates without a copy/instance of the imported/source asset which might have implications or cause inconsistencies in behavior, but difficult to handle when auto-save is enforced.

I had the ambition to refactor the UITK editor to also use `InputActionAssetManager` as part of this PR but gave up on this after partial progress on improving it and adding unit tests for it which at least gets us one step closer to eliminate even more code. This could be resumed in the future as a separate task.

Recommended to cover during quality assurance:
- Copying, renaming, moving and deleting .inputactions assets from within editor with and without editors open.
- Copying, renaming, moving and deleting .inputactions assets from outside editor (OS) with and without editors open.
- Functional test of UITK Input Action Editor with respect to above and general functionality.
- Functional test of IMGUI Input Action Editor with respect to above and general functionality.
- Functional test of UITK Project Settings Input Action Editor with respect to above and general functionality.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests added `Actions_NameIsSetToNameFromJson_IfCreatedFromJsonWithName`, `Actions_NameIsSetToEmptyString_IfCreatedFromJsonWithoutName`, `Actions_NameInJsonIsSetToObjectName_IfCreatedFromObjectWithGivenName()`, `Actions_NameInJsonIsSetToEmptyString_IfCreatedFromObjectWithEmptyName`, `Actions_NameInJsonIsSetToEmptyString_IfCreatedFromObjectWithNullName`, `Editor_InputActionAssetManager_ThrowsIfCreatedFromNullReference`,
`Editor_InputActionAssetManager_ThrowsIfCreatedFromNonPersistedAsset`,
`Editor_InputActionAssetManager_SerializedObjectIsAWorkingCopyOfImportedAsset`,
`Editor_InputActionAssetManager_isDirtyReflectsWhetherWorkingCopyIsDifferentFromImportedAsset`,
`Editor_InputActionAssetManager_CanMoveAssetOnDisk` (modified/simplified),
`Editor_InputActionAssetManager_CanDeleteAssetOnDisk`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

ekcoh committed:
Disabled offending code for further investigation

ekcoh committed:
wip

ekcoh committed:
wip

ekcoh committed:
Code cleanup

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

# Conflicts:
#	Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs
#	Packages/com.unity.inputsystem/InputSystem/Editor/AssetImporter/InputActionImporter.cs

ekcoh committed:
minor edit post merge

ekcoh committed:
FIX: Fixed a regression in IMGUI editor where it would auto-save on focus lost even with toggle disabled.

ekcoh committed:
wip

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

ekcoh committed:
Removed modern code to allow 2019 compilation.

ekcoh committed:
Allow 2019 compilation

ekcoh committed:
wip

ekcoh committed:
piw

Refactoring

ekcoh committed:
feReReR

Refactoring and formatting fix

ekcoh committed:
Code cleanup

ekcoh committed:
Removed unnecessary type constraint

ekcoh committed:
FIX: Added mocking support for all Input System user dialogs and default mocking and restore of dialogs as part of InputTestFixture

ekcoh committed:
wip

ekcoh committed:
Removed unused namespaces

ekcoh committed:
FIX: Assets now properly respond to and update working copy of serializedobject with preserved selection indices.

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

Pauliusd01 commented:
Entering Play Mode in main repo project throws this error (scene does not matter): 
`'UnityEngine.InputSystem.Editor.EnableUITKEditor' is missing the class attribute 'ExtensionOfNativeClass'!`

Pauliusd01 commented:
Trying to add new control schemes throws errors

https://github.com/Unity-Technologies/InputSystem/assets/54306142/f55ae476-26c7-48b0-b094-640144fe5b28

ekcoh commented:
> Trying to add new control schemes throws errors
> 
>  Unity_2024-02-28_14-10-23.mp4

ekcoh committed:
FIX: Removed hard-coded coupling between importer and editor. Removed redundant code in IMGUI editor which is handled via common processor callbacks.

ekcoh commented:
> ExtensionOfNativeClass

I am not getting this error. How did you manage to get this? (Running 2023.3.b04 with cleaned Library folder and checkout)

Pauliusd01 commented:
> > ExtensionOfNativeClass
> 
> I am not getting this error. How did you manage to get this? (Running 2023.3.b04 with cleaned Library folder and checkout)

I don't see it on latest commit either, will let you know if I ever get it again

ekcoh commented:
> Trying to add new control schemes throws errors
> 
>  Unity_2024-02-28_14-10-23.mp4

Found why this is happening. Basically range validation should happen at construction time since I now allow swapping out the serializedobject, fix in the works.

Pauliusd01 commented:
Nullrefs thrown when renaming an already opened input actions asset. (On Develop you don't get these types of errors but you do get a FileNotFoundException instead)
https://github.com/Unity-Technologies/InputSystem/assets/54306142/d4e46c7a-f840-479f-932e-e5e8ca2bb665

ekcoh committed:
FIX: Adressed an issue where lack of index to GUID and GUID to index transformations could lead to invalid selections when switching serializedObject instances in editor.

ekcoh commented:
> > Trying to add new control schemes throws errors
> > Unity_2024-02-28_14-10-23.mp4
> 
> Found why this is happening. Basically range validation should happen at construction time since I now allow swapping out the serializedobject, fix in the works.

@Pauliusd01 this should now be fixed with this commit https://github.com/Unity-Technologies/InputSystem/pull/1850/commits/3215aaa813fa45d9d66d438aa61f1108554b826b along with improved selection index transformation index->GUID and then GUID->index between asset instances when reimported from disc with editor open. This also triggers on save since disc write will trigger reimport so to the editor save and external file changes are similar.

ekcoh commented:
> Nullrefs thrown when renaming an already opened input actions asset. (On Develop you don't get these types of errors but you do get a FileNotFoundException instead)
> https://github.com/Unity-Technologies/InputSystem/assets/54306142/d4e46c7a-f840-479f-932e-e5e8ca2bb665

This was one of my key test cases just before putting this PR in review so must be some mistake I did in last clean-up of the branch. Investigating....

ekcoh committed:
FIX: Desubscribe from state change callbacks in case object destroy or GC is delayed. Fixed UITK editor not showing correct state after a domain reload with editor open. Simplified editor window code. Improved exception safety in editor window to be more robust and close window in case of fatal exception.

ekcoh commented:
> > Nullrefs thrown when renaming an already opened input actions asset. (On Develop you don't get these types of errors but you do get a FileNotFoundException instead)
> > https://github.com/Unity-Technologies/InputSystem/assets/54306142/d4e46c7a-f840-479f-932e-e5e8ca2bb665
> 
> This was one of my key test cases just before putting this PR in review so must be some mistake I did in last clean-up of the branch. Investigating....

This should be fixed in https://github.com/Unity-Technologies/InputSystem/pull/1850/commits/9e4dc181d0eec337e0aba540c87a1df70122887f. Also simplified editor code a bit and fixed an issue where editor would indicate incorrect modification/dirty flag (*) if editor is open during domain reload.

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

ekcoh committed:
Updated CHANGELOG

ekcoh committed:
Merge branch 'isx-1887-asset-being-wiped' of github.com:Unity-Technologies/InputSystem into isx-1887-asset-being-wiped

ekcoh commented:
For visibility I am currently looking into the test failures here which do not reproduce for me locally inside Unity but seem to reproduce here on CI

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Actions.cs:
```diff
@@ -4980,6 +4980,80 @@ public void Actions_CanConvertAssetToAndFromJson()
+    static string MinimalJson(string name = null)
+    {
+        if (name != null)
+            return "{\n    \"name\": \"" + name + "\",\n    \"maps\": [],\n    \"controlSchemes\": []\n}";
```
I find this a bit hard to parse, perhaps using string literal might help

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated?redirectedfrom=MSDN#interpolated-raw-string-literals

E.g.

string s = $"""
    "name": {name},
    "maps": [],
	"controlSchemes": []
"""

For some reason this didn't work for me

string s = $"""
{{
    "name": {name},
    "maps": [],
	"controlSchemes": []
}}
"""

ekcoh replied:
I tried that but ended up with so much escaping it was even more unreadable. Will try your provided sample though. The benefit with this one is that its comparable to the formatting generated by the JSON parser. I guess another alternative is to make it readable, convert an object from it, then back to json and store that as a "constant" to be used later.

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Actions.cs:
```diff
@@ -4980,6 +4980,80 @@ public void Actions_CanConvertAssetToAndFromJson()
+    public void Actions_NameInJsonIsSetToEmptyString_IfCreatedFromObjectWithNullName()
+    {
+        // When serializing JSON from object without a given name, Unity forces the name to empty string.
+        // Basically Unity prevents us from doing serialization an omit optional members in a convenient way.
```
Text here looks slightly off

Should the **an** be **and**:
Basically Unity prevents us from doing serialization **and** omit**s** optional members in a convenient way.

ekcoh replied:
Thanks, fixed in https://github.com/Unity-Technologies/InputSystem/pull/1850/commits/e8fb5dd4c1b6870163e809ae8bdd15e76270be85

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor_InputActionAssetManager.cs:
```diff
@@ -0,0 +1,134 @@
+    {
+        const string filename = "InputAsset." + InputActionAsset.Extension;
+        var directoryBeforeMove = AssetDatabaseUtils.CreateDirectory();
+        var directoryAfterMove = AssetDatabaseUtils.RandomDirectoryPath();
```
Why switched to random path here.
Probably fine - but seems unnecessary refactoring for this pr

ekcoh replied:
Switching to random paths was not strictly necessary but the reason why I introduced that utility is the asset file mess that currently exists in our tests leaving all sorts of files as left overs. Hence why I created AssetDatabaseUtil.cs. Random paths may be annoying but originated as that to avoid all the clashes test cases caused when not removing things properly in teardown. The added util takes care of this and guarantees that there are never clashes even if file cleanup would fail.

ekcoh replied:
My vision with it is to reuse in other tests to gradually clean this up down the road. Used it here since these tests were contributing to leaving test files and directories generating warnings in console after test run and leaving a dirty asset folder after test run.

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor_InputActionAssetManager.cs:
```diff
@@ -0,0 +1,134 @@
+            inputActionAssetManager.Initialize();
+
+            FileUtil.MoveFileOrDirectory(directoryBeforeMove, directoryAfterMove); // TODO Wouldn't move .meta files
+            FileUtil.MoveFileOrDirectory(directoryBeforeMove + ".meta", directoryAfterMove + ".meta");
```
Original test didn't move the meta file - not sure why that's changed either

ekcoh replied:
This is true, and this resulted in a warning, something like "XXX.meta file not removed". I changed this to move them in pairs instead to mimic what you should do in file system and what AssetDatabase does under the hood. The above is now a single method AssetDatabaseUtil.ExternalMoveFileOrDirectory. A corresponding exist using ADB instead.

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor.cs:
```diff
@@ -834,36 +831,6 @@ public void Editor_InputActions_AssetFileUpdatedAfterMoveOrCopy(string newAssetP
-
-        var inputActionAssetManager = new InputActionAssetManager(asset);
-        inputActionAssetManager.Initialize();
-        inputActionAssetManager.onDirtyChanged = (bool dirty) => {};
```
This is removed when refactoring. Not sure why this was present - perhaps it was previously here to avoid any dialog boxes ?

ekcoh replied:
Yes, it was an odd design I modified since it didn't make sense to me and provided a somewhat messy API. Basically when this type is constructed via a constructor it doesn't make sense to require a call to Initialize after. This is now done as part of the constructor which is cleaner in my opinion. Initialize() remains available for serialisation scenarios when it should be called, but then no constructor is called.
The other odd thing here was that SaveChangesToAsset() would suprisingly throw if onDirtyChanged callback was not set. This didn't sit well I thought so I made it optional to set a callback and if you have registered one you get a callback on Save (transition to false), while its also fine to save without a callback.

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor_InputActionAssetManager.cs:
```diff
@@ -0,0 +1,134 @@
+
+            inputActionAssetManager.SaveChangesToAsset();
+
+            var fileContents = File.ReadAllText(AssetDatabase.GetAssetPath(asset));
```
Previously this was checking the target path specifically - we now no longer check that so potentially its not been moved.
May be safe but there seems to be more refactoring that I would expect in this test.

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor.cs:
```diff
@@ -834,36 +831,6 @@ public void Editor_InputActions_AssetFileUpdatedAfterMoveOrCopy(string newAssetP
-        var fileContents = File.ReadAllText(kAssetPathAfterMove);
-        Assert.AreNotEqual(kDefaultContents, fileContents, "Expected file contents to have been modified after SaveChangesToAsset was called.");
-
-        AssetDatabase.DeleteAsset("Assets/DirectoryAfterRename");
```
This clean up is removed from the test - so doesn't it now leak the asset ?

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor_InputActionAssetManager.cs:
```diff
@@ -0,0 +1,134 @@
+            var fileContents = File.ReadAllText(AssetDatabase.GetAssetPath(asset));
+            Assert.AreNotEqual(kDefaultContents, fileContents, "Expected file contents to have been modified after SaveChangesToAsset was called.");
+        }
+    }
```
No clean up of asset so is this now leaking the test?

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor_InputActionAssetManager.cs:
```diff
@@ -0,0 +1,134 @@
+    public void Editor_InputActionAssetManager_SerializedObjectIsAWorkingCopyOfImportedAsset()
+    {
+        var asset = AssetDatabaseUtils.CreateAsset<InputActionAsset>();
+        using (var sut = new InputActionAssetManager(asset))
```
sut is a bit of an odd variable name.
I don't know what that actually is supposed to mean

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_Editor_InputActionAssetManager.cs:
```diff
@@ -0,0 +1,134 @@
+            // Add a map to editable object and apply changes, expect it to not be dirty since we haven't explicitly updated dirty state
+            var mapProperty = InputActionSerializationHelpers.AddActionMap(sut.serializedObject);
+            sut.ApplyChanges();
+            Assert.That(sut.dirty, Is.False); // TODO Note: dirty flag is not updated only by applying changes
```
Is this is expected then I don't know why the TODO is present

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_ProjectWideActions.cs:
```diff
@@ -29,8 +29,8 @@ internal partial class CoreTests
-    const string m_AssetBackupDirectory = "Assets/~TestBackupFiles";
-    const string s_DefaultProjectWideAssetBackupPath = "Assets/~TestBackupFilesDefaultProjectWideAssetBackup.json";
+    const string kAssetBackupDirectory = "Assets/~TestBackupFiles";
+    const string kDefaultProjectWideAssetBackupPath = kAssetBackupDirectory + "/DefaultProjectWideAssetBackup.json";
```
Looks like a missing / was added here. I think thats intentional

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_ProjectWideActions.cs:
```diff
@@ -44,10 +44,10 @@ public void OneTimeSetUp()
-            if (!Directory.Exists(m_AssetBackupDirectory))
-                Directory.CreateDirectory(m_AssetBackupDirectory);
+            if (!Directory.Exists(kAssetBackupDirectory))
+                AssetDatabase.CreateFolder("Assets", "~TestBackupFiles");
```
Why the change from CreateDirectory to CreateFolder ?
If a user did it out side the test via folder browser the wouldn't be using AssetDatabase API.

lyndon-unity commented on Assets/Tests/InputSystem/CoreTests_ProjectWideActions.cs:
```diff
@@ -57,14 +57,13 @@ public void OneTimeTearDown()
                 newPath: ProjectWideActionsAsset.defaultAssetPath);
-            Directory.Delete("Assets/~TestBackupFiles");
-            File.Delete("Assets/~TestBackupFiles.meta");
+            AssetDatabase.DeleteAsset(kAssetBackupDirectory);
```
Probably fine but not sure why changed - seems more change than needed in this PR.

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Actions/InputActionAsset.cs:
```diff
@@ -1,7 +1,6 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
-using System.Linq;
```
Nice removal :)

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Actions/InputActionAsset.cs:
```diff
@@ -297,14 +296,12 @@ public IEnumerable<InputBinding> bindings
         public string ToJson()
         {
-            var fileJson = new WriteFileJson
+            return JsonUtility.ToJson(new WriteFileJson
```
Is this just a style choice or are you doing this for performance ?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Actions/InputActionAsset.cs:
```diff
@@ -948,6 +945,13 @@ internal struct WriteFileJson
             public InputControlScheme.SchemeJson[] controlSchemes;
         }
 
+        [Serializable]
```
Doesn't seem used unless I'm missing something

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs:
```diff
@@ -1,5 +1,6 @@
 #if UNITY_EDITOR
 using System;
+using System.Diagnostics.CodeAnalysis;
```
Is this needed?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs:
```diff
@@ -51,20 +63,29 @@ private InputActionAsset importedAsset
             }
         }
 
+        public InputActionAsset editedAsset => m_AssetObjectForEditing; // TODO Remove if redundant
```
Why added if then also asking if should be removed ?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs:
```diff
@@ -51,20 +63,29 @@ private InputActionAsset importedAsset
+                throw new Exception($"Failed to get asset {inputActionAsset.name} GUID");
+
             m_ImportedAssetObject = inputActionAsset;
-            bool isGUIDObtained = AssetDatabase.TryGetGUIDAndLocalFileIdentifier(importedAsset, out m_AssetGUID, out long _);
```
Seem to be deciding to change from warning to Exception. Not sure why though

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs:
```diff
@@ -91,45 +114,65 @@ public bool Initialize()
         {
+            if (m_SerializedObject == null)
+                return;
             m_SerializedObject?.Dispose();
```
No longer need ?. if checking above

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs:
```diff
@@ -91,45 +114,65 @@ public bool Initialize()
         }
 
-        private void CreateWorkingCopyAsset()
+        public static InputActionAsset CreateWorkingCopy(InputActionAsset source)
```
Do these need to be public - I don't see them called outside the file

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/Internal/EditorHelpers.cs:
```diff
@@ -60,31 +69,61 @@ public static bool CheckOut(string path)
+                return false;
+
+            // (Over)write file text content.
+            File.WriteAllText(GetPhysicalPath(assetPath), text);
```
I don't really understand what the refactoring solves here
EditorHelpers.WriteAllText -> File.WriteAllText

Seems ok, but not clear why needed

ekcoh committed:
Refactored InputAssetManager tests out of edit/play mode assembly and into edit mode only assembly since they doesn't make sense to execute in play mode. Also moved related utils.

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs:
```diff
@@ -191,44 +219,30 @@ internal static bool SaveAsset(string assetPath, string assetJson)
+                return true;
 
-            return true;
+            Debug.LogError($"Unable save asset to \"{assetPath}\" since the asset-path could not be checked-out as editable in the underlying version-control system.");
```
Logic change to remove ! seems ok but not really sure why needed in this PR.
A few changes seem more than necessary for focused fixes.
I wouldn't block the PR for it but it makes it harder to review and seems a style choice.

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

ekcoh committed:
Fixed merge mistake and reverted accidental class rename from previously.

ekcoh commented:
> For visibility I am currently looking into the test failures here which do not reproduce for me locally inside Unity but seem to reproduce here on CI

I noticed the asset manager tests where running in player/play mode which make little sense so moved them to editor assembly. Will see if this helps the problem since only failed in play mode (likely to sandbox nature of it mixed with ADB and file operations).

Update: Above didn't work but noticed only Windows failures and native Windows paths to added a fix to convert to Unix path separators when processing paths in CreateDirectories. New attempt to fix test errors in https://github.com/Unity-Technologies/InputSystem/pull/1850/commits/f9740ab8f9f76c85baf0a75c2cd0e15b56df66be

ekcoh committed:
FIX: Handle Windows native paths in CreateDirectories of AssetDatabaseUtils.cs

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionEditorWindow.cs:
```diff
@@ -263,7 +246,7 @@ private void SetAsset(InputActionAsset asset)
 
             m_ActionAssetManager = new InputActionAssetManager(asset) {onDirtyChanged = OnDirtyChanged};
-            m_ActionAssetManager.Initialize();
+            //m_ActionAssetManager.Initialize(); // TODO No longer needed when using constructor
```
Seems like this should be deleted ?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionEditorWindow.cs:
```diff
@@ -854,64 +838,22 @@ public void Dispose()
         private bool m_ForceQuit;
 
-        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1812:AvoidUninstantiatedInternalClasses", Justification = "Intantiated through reflection by Unity")]
-        private class ProcessAssetModifications : UnityEditor.AssetModificationProcessor
```
So this is moved to IInputActionAssetEditor

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetImporter/InputActionImporter.cs:
```diff
@@ -331,38 +319,72 @@ public static void CreateInputAsset()
-        // hold the old name and won't match what's in memory (asset looks "dirty"). To work around this, we
-        // must flush the updated JSON to the file whenever this operation occurs.
+        // File extension of the associated asset
+        private const string kFileExtension = "." + InputActionAsset.Extension;
```
Seems reasonable adding null check but the extra movement in the file made it harder to diff.

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetImporter/InputActionImporter.cs:
```diff
@@ -58,7 +50,7 @@ private static InputActionAsset CreateFromJson(AssetImportContext context)
             string content;
             try
             {
-                content = EditorHelpers.ReadAllText(context.assetPath);
```
Some of this extra refactoring seemed unnecessary for this pr - tidy maybe but had made the pr longer to review

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/Internal/EditorHelpers.cs:
```diff
@@ -1,5 +1,6 @@
 #if UNITY_EDITOR
 using System;
+using System.Diagnostics.CodeAnalysis;
```
Is this needed?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:
```diff
@@ -483,7 +483,9 @@ public static Command SaveAsset(Action postSaveAction)
-                InputActionAssetManager.SaveAsset(state.serializedObject.targetObject as InputActionAsset);
+                // TODO This needs to callback to owning editor to save or have asset GUID
+                // TODO It makes more sense to call back to editor since editor owns target object?
+                //InputActionAssetManager.SaveAsset(state.serializedObject.targetObject as InputActionAsset);
```
Can we clean up the TODO here
Delete this now ? If SaveAction is handing it ?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:
```diff
@@ -498,7 +500,7 @@ public static Command ToggleAutoSave(bool newValue, Action postSaveAction)
                     if (newValue)
                     {
-                        InputActionAssetManager.SaveAsset(state.serializedObject.targetObject as InputActionAsset);
+                        //InputActionAssetManager.SaveAsset(state.serializedObject.targetObject as InputActionAsset);
```
Delete this now ? If SaveAction is handing it ?

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/InputActionsEditorView.cs:
```diff
@@ -52,6 +55,16 @@ public InputActionsEditorView(VisualElement root, StateContainer stateContainer,
             autoSaveToggle.value = InputEditorUserSettings.autoSaveInputActionAssets;
             autoSaveToggle.RegisterValueChangedCallback(OnAutoSaveToggle);
 
+            // Hide save toolbar if there is no save action provided since we cannot support it
```
Fine to move this but didn't seem like a critical refactoring

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorState.cs:
```diff
@@ -14,12 +14,12 @@ internal struct InputActionsEditorState
         public int selectedControlSchemeIndex { get { return m_selectedControlSchemeIndex; } }
         public int selectedDeviceRequirementIndex { get  {return m_selectedDeviceRequirementIndex; } }
-        public InputControlScheme selectedControlScheme => m_ControlScheme;
+        public InputControlScheme selectedControlScheme => m_ControlScheme; // TODO Bad this either po
```
TODO seems to have missing text

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorState.cs:
```diff
@@ -54,34 +56,94 @@ internal struct InputActionsEditorState
+                    m_selectedControlSchemeIndex = 0;
+                m_ControlScheme = new InputControlScheme(
+                    controlSchemesArrayProperty.GetArrayElementAtIndex(other.m_selectedControlSchemeIndex));
+                // TODO Preserve device requirement index
```
Incomplete TODO - not sure if this needs doing before landing. Seems no worse than it was

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorWindow.cs:
```diff
@@ -123,67 +124,86 @@ private static InputActionsEditorWindow GetOrCreateWindow(int id, out bool isAlr
         private void SetAsset(InputActionAsset asset, string actionToSelect = null, string actionMapToSelect = null)
         {
-            m_AssetPath = AssetDatabase.GetAssetPath(asset);
-            var serializedAsset = new SerializedObject(asset);
```
ok looks like this moved to TryUpdateFromAsset

ekcoh replied:
Correct. Why repeat the same code in multiple places.

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorWindow.cs:
```diff
@@ -260,34 +271,48 @@ private void OnLostFocus()
+                case Dialog.Result.Discard:
                     // Don't save, quit - reload the old asset from the json to prevent the asset from being dirtied
-                    AssetDatabase.ImportAsset(m_AssetPath);
+                    // TODO I doubt this is needed since we were operating on a copy?!
```
Delete this then ?

ekcoh replied:
Removed in https://github.com/Unity-Technologies/InputSystem/pull/1850/commits/a8c93599cbc79b3f13fa0a0cad68d20483954d32

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorWindowUtils.cs:
```diff
@@ -13,18 +15,35 @@ internal class InputActionsEditorWindowUtils
-        public static ConfirmSaveChangesDialogResult ConfirmSaveChanges(string path)
+        // Similar to InputActionAsset.ToJson() but converts to JSON excluding the name property and any additional JSON
+        // content that may be part of the file not recognized as required data.
+        public static string ToJsonWithoutName(InputActionAsset asset)
```
Are these actually used and do we really need them ?

ekcoh replied:
Yes they are used to convert an asset to JSON but exclude the name field. It is used for comparing two assets content to update dirty flags. @timkeo made a change previously which aligns JSON and SerializedObject names when desynchronised, but there is no need for editor to go dirty if that is the only change, similar to us discarding other "augmented" fields in json that users might enter. This also allowed removing a previous workaround for comparing whether an asset was empty or not that was a bit clunky.

ekcoh replied:
I noticed now however that the Read version is not used anymore (refactored that one away) so I removed it in a coming commit.

lyndon-unity commented on Packages/com.unity.inputsystem/Tests/TestFixture/InputTestFixture.cs:
```diff
@@ -140,6 +140,13 @@ public virtual void Setup()
                 InputSystem.settings.SetInternalFeatureFlag(InputFeatureNames.kUseReadValueCaching, true);
                 InputSystem.settings.SetInternalFeatureFlag(InputFeatureNames.kParanoidReadValueCachingChecks, true);
+
+                // Default mock dialogs to avoid unexpected cancellation of standard flows
```
Handy to keep testing flowing

ekcoh replied:
Or to set them and actually test the dialogs in the future which currently has zero test coverage from auto tests.

lyndon-unity commented on Packages/com.unity.inputsystem/Tests/TestFixture/ScopedDisposable.cs:
```diff
@@ -0,0 +1,57 @@
+    // Utility allowing access to object T as well as a dispose delegate to clean-up any resources associated with it.
+    // Useful with the 'using' keyword to provide scoped RAII-like cleanup of objects in tests without having a
+    // dedicated fixture handling the clean-up.
+    public sealed class ScopedDisposable<T> : IDisposable
```
This doesn't seem to be used unless I'm missing something

ekcoh replied:
Correct, it was initially used but I managed to replace all its use with AssetDatabaseUtil.cs. I would like to keep it for now if OK since there are many use-cases for it in existing ill-formed tests that leak resources if any Assertion fails.

ekcoh replied:
It works really nice with C# 8 "using var", but doesn't seem to work well with Unity 2019 then... :/ it will be usable as soon as modern C# can be used in the tests since "using var scoped = Scoped.CreateObject<T>();" would auto clean-up regardless of where test execution was aborted and avoids writing tons of TearDown functions to handle it for each different scenario.

lyndon-unity reviewed (commented):
Seems ok - few comments about some todo's that can be removed.
Overall seems to be better, but couple of failing tests need fixing.

ekcoh committed:
FIX: Updated AssetDatabaseUtils to always return UNIX style paths and avoid mixing directory separator characters. This makes it easier to use for assertions but also compliance across platforms. Intention is to resolve test issues on Windows on this PR. Also restored assertion in test accidentally removed.

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

ekcoh committed:
Cleanup based on review feedback

ekcoh commented:
> Seems ok - few comments about some todo's that can be removed.
> Overall seems to be better, but couple of failing tests need fixing.

Seems like the update on test utility and tests made those tests pass now. Other CI issues (also seen on other branches remain)

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

Pauliusd01 commented:
Is it okay to get this "failed to delete" prompt? Cancel should cancel deletion instead of letting it fail. Although this does not seem to break anything so I'm fine with keeping it as is, if It's not worth it

https://github.com/Unity-Technologies/InputSystem/assets/54306142/ac77f253-9142-4a96-abee-3d175afbdfea

Pauliusd01 reviewed (approved):
LGTM. Checked deleting and renaming input assets while they're open in different ways, checked basic input functionality in Editor/Player and checked Input Action Window functionality after Editor restarts (This PR also fixes ISX-1904 so I've closed that one)

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs:
```diff
@@ -33,12 +41,16 @@ public string path
             {
-                if (m_ImportedAssetObject != null)
-                    return m_ImportedAssetObject.name;
+                var asset = importedAsset;
```
looks like we could skip creating var asset here.

ekcoh replied:
I guess you are right, its kind of intentional though since importedAsset when not just cached is a costly operation. Hence I keep a local to avoid any side-effects or cost introduced by calling importedAsset multiple times. Since there are many access to asset variable in this getter though and importedAsset is non-trivial I rather just keep the local. Otherwise if someone is inventive down the line with importedAsset this getter could be causing problems.

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/Dialog.cs:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        // User UI dialog windows related to InputActionAssets
+        public static class InputActionAsset
```
naming: InputActionAssetDialog?

ekcoh replied:
Am I understanding correctly that you would you prefer `Dialog.InputActionAssetDialog.ShowSaveChanges(path)` in favour of `Dialog.InputActionAsset.ShowSaveChanges(path)`?  
Since already prefixed and scope reduced by outer `Dialog` class my intention was to avoid excessive long names and also avoid scope leakage into outer scope which happens when you just use namespace and auto-import. The idea here was to access Dialog since that is what you want to use and then for what kind of type and then use-case. I can change but just want to know if I understand correctly.

ritamerkl replied:
I think it's good to avoid any confusion with the names in all scenarios, if we are forced to always use Dialog.xx I think this is sufficient.

ekcoh replied:
ok will leave it as is for now then and we can reconsider in the future

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/Dialog.cs:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        // User UI dialog windows related to InputControlSchemes
+        public static class ControlScheme
```
naming: ControlSchemeDialog?

ekcoh replied:
Would fall under same comment/question as above.

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorState.cs:
```diff
@@ -242,18 +304,32 @@ private HashSet<int> GetOrCreateExpandedStateInternal()
         }
 
-        private readonly Dictionary<(string, string), HashSet<int>> m_ExpandedCompositeBindings;
+        private static SerializedProperty GetSelectedAction(SerializedProperty map, int actionIndex)
```
Selectors has this method already.

ekcoh replied:
Yes but only using a InputActionState and it would be bad practise to pass this before object is fully constructed. I guess we could refactor Selectors to have to GetSelectedAction functions and also refactor the ActionMap ones to accept SerializedObject as an overload...

ritamerkl replied:
Your suggestion sounds reasonable to me.

ekcoh replied:
Fixed in https://github.com/Unity-Technologies/InputSystem/pull/1850/commits/7541e17d9342e40e515f3f7637b3160ca139aeaf
Also found another method I could remove in this class

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/InputActionsEditorView.cs:
```diff
@@ -221,13 +234,23 @@ public class ViewState
+        {
+            base.DestroyView();
+
+            Debug.Log("Destroy View");
```
Maybe remove method? looks like this is redundant and just for debugging purpose.

ekcoh replied:
This should be removed yes, will fix

ekcoh replied:
Fixed in https://github.com/Unity-Technologies/InputSystem/pull/1850/commits/7541e17d9342e40e515f3f7637b3160ca139aeaf

ritamerkl reviewed (approved):
LGTM. The comments and explanation blocks are helpful and a good inspiration for documenting code. Added some minor things to consider.

ekcoh commented:
> LGTM. Checked deleting and renaming input assets while they're open in different ways, checked basic input functionality in Editor/Player and checked Input Action Window functionality after Editor restarts (This PR also fixes ISX-1904 so I've closed that one)

Awesome, always happy to have some positive side-effects on the open bug count :)

ekcoh commented:
> LGTM. The comments and explanation blocks are helpful and a good inspiration for documenting code. Added some minor things to consider.

Happy to hear that if it helped readability of the code. Just don't take the size and scattered focus of this PR as inspiration since its bad practise to create such a big PR covering too much ground.

ekcoh committed:
Address review feedback: Removed debug code, removed code and replaced with Selector usage.

ekcoh committed:
Merge branch 'develop' into isx-1887-asset-being-wiped

# Conflicts:
#	Packages/com.unity.inputsystem/CHANGELOG.md

ekcoh committed:
Attempt to fix formatting problem on CHANGELOG.md
